src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -73,22 +69,19 @@ public static object Deserialize(string json, Type returnType, JsonSerializerOpt
             object result;
             byte[] tempArray = null;
 
-            int maxBytes;
-
-            // For performance, avoid asking for the actual byte count unless necessary.
-            if (json.Length > MaxArrayLengthBeforeCalculatingSize)
+            // For performance, avoid obtaining the actual byte count unless the memory usage may be
+            // higher than the threshold which is options.DefaultBufferSize.
+            long maxBytes = (long)json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding;
+            if (maxBytes > options.DefaultBufferSize)
             {
                 // Get the actual byte count in order to handle large input.
                 maxBytes = JsonReaderHelper.GetUtf8ByteCount(json.AsSpan());
```

Comment:
nit: Given this throws for certain large string lengths (when the return count won't fit int.MaxValue), should this be documented as an exception on the method?

Reply:
If so we should also document every other Deserialize method as well. Perhaps create a doc issue?

Reply:
Do the other deserialize methods have this transcoding-based exception though? If so, then yes, we should.